### PR TITLE
feat: @wdio/xvfb

### PR DESCRIPTION
**Pull request by goosewobbler**
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

Creating a new package `@wdio/xvfb` for running XVFB on linux systems where appropriate to avoid confusing Chromium errors.  See https://github.com/webdriverio/webdriverio/issues/14501

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Polish (an improvement for an existing feature)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (improvements to the project's docs)
- [ ] Specification changes (updates to WebDriver command specifications)
- [ ] Internal updates (everything related to internal scripts, governance documentation and CI files)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added the necessary documentation (if appropriate)
- [x] I have added proper type definitions for new commands (if appropriate)

## Backport Request

[//]: # (The current `main` branch is the development branch for WebdriverIO v9. If your change should be released to the current major version of WebdriverIO (v8), please raise another PR with the same changes against the `v8` branch.)

- [ ] This change is solely for `v9` and doesn't need to be back-ported
- [ ] Back-ported PR at `#XXXXX`

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

**Comment by goosewobbler** [comment]
Marked as ready for review for some early feedback.  Still need to consider the following, I think we need manual tests at a minimum:

* E2E / Smoke tests
* Manual tests
* Docs updates

Specifically need to verify that (1) xvfb is being run on a system where it is available, and (2) xvfb is installed and run on a system where it is not available.  These should be verified on different Linux distros.

File: packages/wdio-local-runner/src/index.ts [hunk 1]
```diff
@@ -16,19 +18,32 @@ export interface RunArgs extends Workers.WorkerRunPayload {
 
 export default class LocalRunner {
     workerPool: Record<string, WorkerInstance> = {}
+    private processFactory: ProcessFactory
```
Comment (christian-bromann) [inline packages/wdio-local-runner/src/index.ts]: Unless we explicitly need access to it in a unit test why not just do:


```suggestion
    #processFactory = new ProcessFactory(xvfb)
```

File: packages/wdio-xvfb/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        const isCI = !!(
+            process.env.CI ||
+            process.env.GITHUB_ACTIONS ||
+            process.env.JENKINS_URL ||
+            process.env.TRAVIS ||
+            process.env.CIRCLECI ||
+            process.env.GITLAB_CI ||
+            process.env.BUILDKITE ||
+            process.env.APPVEYOR
+        )
```
Comment (christian-bromann) [inline packages/wdio-xvfb/src/index.ts]: Maybe we should just use https://www.npmjs.com/package/is-ci or similar?

File: packages/wdio-xvfb/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                                    return 'alpine'
+                                } catch {
+                                    return 'unknown'
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
```
Comment (christian-bromann) [inline packages/wdio-xvfb/src/index.ts]:
```suggestion
            // Fallback detection methods
            const dist = await Promise.all([
              execAsync('which apt-get').then(() => 'debian', () => {}),
              execAsync('which yum').then(() => 'rhel', () => {}),
              execAsync('which dnf').then(() => 'fedora', () => {}),
              execAsync('which zypper').then(() => 'suse', () => {}),
              execAsync('which pacman').then(() => 'arch', () => {}),
              execAsync('which apk').then(() => 'alpine', () => {})
            ])
            return dist.unshift() || 'unknown'
        }
```

Comment (christian-bromann) [reply packages/wdio-xvfb/src/index.ts]: Maybe there is an NPM package for this as well?

File: packages/wdio-local-runner/src/processFactory.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+                '--',
+                'node',
+                ...nodeArgs
+            ], {
+                cwd,
+                env,
+                stdio
+            } as SpawnOptions)
+        }
+        log.debug('Creating worker process with regular fork')
```
Comment (christian-bromann) [inline packages/wdio-local-runner/src/processFactory.ts]: Is this something we can include the xvfb package?

File: packages/wdio-xvfb/README.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    
+    await xvfb.stop()
+}
+```
+
+### Browser Testing Setup
+
+```js
+import { xvfb } from '@wdio/xvfb'
+import { Builder } from 'selenium-webdriver'
```
Comment (christian-bromann) [inline packages/wdio-xvfb/README.md]: This should not be docs about Selenium 😉

File: packages/wdio-xvfb/LICENSE [hunk 6]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2024 WebdriverIO
```
Comment (christian-bromann) [inline packages/wdio-xvfb/LICENSE]:
```suggestion
Copyright (c) 2025 OpenJS Foundation
```

File: packages/wdio-xvfb/src/index.ts (hunk 2 above)
Comment (goosewobbler) [reply packages/wdio-xvfb/src/index.ts]: Hmmm thought I already had but I think it got lost in an update.

File: packages/wdio-local-runner/src/processFactory.ts (hunk 4 above)
Comment (goosewobbler) [reply packages/wdio-local-runner/src/processFactory.ts]: We could, but I thought it better to keep the process management in the local-runner.  Not too bothered though, it could certainly live in the xvfb package.

File: packages/wdio-xvfb/src/index.ts (hunk 3 above)
Comment (goosewobbler) [reply packages/wdio-xvfb/src/index.ts]: I'll check this.  Definitely need something better than that pyramid.

File: packages/wdio-xvfb/README.md (hunk 5 above)
Comment (goosewobbler) [reply packages/wdio-xvfb/README.md]: Classic Claude.  Yeah this doc needs a good going over.

**Comment by goosewobbler** [comment]
Conceptually I am wondering if these E2Es are more like smoke tests, will see if they fit better there.  Also need to add tests for the other distros, and fix the perm errors.

UPDATE: changed my mind, these don't fit well with the existing smokes.

File: packages/wdio-local-runner/src/processFactory.ts (hunk 4 above)
Comment (christian-bromann) [reply packages/wdio-local-runner/src/processFactory.ts]: Yeah, I think it makes conceptional more sense

File: packages/wdio-xvfb/src/index.ts (hunk 3 above)
Comment (goosewobbler) [reply packages/wdio-xvfb/src/index.ts]: I think rolling our own is better in this case, couldn't find a recently updated package for this.

File: packages/wdio-xvfb/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+        // Only run on Linux systems
+        if (os.platform() !== 'linux') {
+            return false
+        }
+
+        // Check if we're in a headless environment (no DISPLAY set or in CI)
+        const hasDisplay = process.env.DISPLAY
+        return !hasDisplay || isCI
```
Comment (goosewobbler) [inline packages/wdio-xvfb/src/index.ts]: I wonder if we also want to detect the `--headless` Chromium flag in the local runner and force this accordingly?  That's to satisfy the edge case where display is set, not in CI but `--headless` has been explicitly set

**Comment by goosewobbler** [comment]
Still need to go over the docs but I think it's pretty much there.

File: packages/wdio-xvfb/src/index.ts (hunk 7 above)
Comment (goosewobbler) [reply packages/wdio-xvfb/src/index.ts]: Added support for this case.

**Comment by goosewobbler** [comment]
These sporadic unit errors are very annoying: 
https://github.com/webdriverio/webdriverio/actions/runs/16598435779/job/46951240548?pr=14663
https://github.com/webdriverio/webdriverio/actions/runs/16598737987/job/46952327105?pr=14663

**Comment by goosewobbler** [comment]
Still need to go over the docs but I think it's getting there.  So far the only documentation I added is the `@wdio/xvfb` README, do we need to mention this functionality anywhere else?  Perhaps in the [Testrunner pages](https://webdriver.io/docs/testrunner)?

Also, do we want this integrated anywhere else or is it just applicable to the local-runner?   I will update the electron service to use it directly in the script / non-testrunner-based case - see https://github.com/webdriverio-community/wdio-electron-service/issues/1132.

File: packages/wdio-local-runner/src/index.ts [hunk 8]
```diff
@@ -1,40 +1,52 @@
...
 const log = logger('@wdio/local-runner')
 
 export type { WorkerInstance }
+export { ProcessFactory }
 
 export interface RunArgs extends Workers.WorkerRunPayload {
-    command: string
-    args: Workers.WorkerMessageArgs
+    command: string;
+    args: Workers.WorkerMessageArgs;
```
Comment (christian-bromann) [inline packages/wdio-local-runner/src/index.ts]:
```suggestion
    command: string
    args: Workers.WorkerMessageArgs
```

File: packages/wdio-local-runner/src/index.ts [hunk 9]
```diff
@@ -98,5 +148,12 @@ export default class LocalRunner {
                 }
             }, 250)
         })
+
+        // Xvfb cleanup is handled automatically by xvfb-run
+        if (xvfb.shouldRun()) {
+            log.info('Xvfb cleanup handled automatically by xvfb-run')
+        }
```
Comment (christian-bromann) [inline packages/wdio-local-runner/src/index.ts]: Is this logging important to users?
```suggestion
```

File: packages/wdio-types/src/Options.ts [hunk 10]
```diff
@@ -327,6 +327,11 @@ export interface Testrunner extends Hooks, WebdriverIO, WebdriverIO.HookFunction
      * Shard tests and execute only the selected shard. Specify in the one-based form like `{ total: 5, current: 2 }`.
      */
     shard?: ShardOptions
+    /**
+     * Disable automatic Xvfb initialization in local runner.
+     * When enabled, tests need to manually call xvfb.init() if needed.
```
Comment (christian-bromann) [inline packages/wdio-types/src/Options.ts]:
```suggestion
     * When enabled, tests need to manually call `xvfb.init()` if needed.
     * @default false
```

File: packages/wdio-local-runner/src/index.ts [hunk 11]
```diff
@@ -1,40 +1,52 @@
...
 import type { Workers } from '@wdio/types'
 
 import WorkerInstance from './worker.js'
+import { ProcessFactory } from './processFactory.js'
 import { SHUTDOWN_TIMEOUT, BUFFER_OPTIONS } from './constants.js'
 
 const log = logger('@wdio/local-runner')
 
 export type { WorkerInstance }
+export { ProcessFactory }
```
Comment (christian-bromann) [inline packages/wdio-local-runner/src/index.ts]: Do we consume this `ProcessFactory` somewhere outside? I still thing this may be better located in the xvfb package, what do you think?

Comment (goosewobbler) [reply packages/wdio-local-runner/src/index.ts]: I'll see what it looks like...

Comment (goosewobbler) [reply packages/wdio-local-runner/src/index.ts]: I think you were right, it's better...

**Comment by goosewobbler** [comment]
Note to self: verify that the case where xvfb fails to start is handled gracefully.  I think some configurable number of retries would make sense, as I have been experiencing occasional failing jobs on Linux CI in this regard (using xvfb-maybe).

https://github.com/webdriverio-community/wdio-electron-service/actions/runs/16750196929/job/47418464616?pr=1133

UPDATE:

prototyping this approach in https://github.com/webdriverio-community/wdio-electron-service/pull/1133...

Seems to work: https://github.com/webdriverio-community/wdio-electron-service/actions/runs/16756083118/job/47439022995?pr=1133
